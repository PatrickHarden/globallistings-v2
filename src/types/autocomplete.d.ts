import { Dictionary } from './common/dictionary';
import { Property } from './property';

// TODO remove the extra properties (except for isProperty) and move them into a property search interface to store in state
interface Autocomplete {
  readonly placeId: string;
  readonly description: string;
  readonly propertyResult?: PropertyResult;
}

interface PropertyResult {
  readonly propertyPartialPath?: string;
  readonly propertyAspects?: string[];
  readonly searchPath?: string;
}

interface NormalisedAutocompleteResult {
  readonly entities: AutocompleteEntities;
  readonly result: string[];
}

interface AutocompleteEntities {
  readonly suggestions: Dictionary<Autocomplete>;
}
