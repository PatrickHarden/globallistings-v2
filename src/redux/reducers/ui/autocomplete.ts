import { ActionType } from 'typesafe-actions';
import { suggestionsSuccess } from '../../actions/autocomplete-request';
import { selectedAutocompletionItem } from '../../actions/autocomplete';

import { DeepPartial } from 'redux';
import { Autocomplete } from '../../../types/ui';

export type GeocodeActions = ActionType<
  typeof suggestionsSuccess | typeof selectedAutocompletionItem
>;

export default (
  state: DeepPartial<Autocomplete> = {},
  action: GeocodeActions
): DeepPartial<Autocomplete> => {
  switch (action.type) {
    case 'GET_SUGGESTIONS_SUCCESS':
      return { ...state, searchTerm: action.meta.hash };
    case 'SELECT_AUTOCOMPLETION_ITEM':
      return { ...state, selectedSuggestionId: action.payload };
    default:
      return state;
  }
};
