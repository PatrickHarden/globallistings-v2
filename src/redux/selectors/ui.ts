import { createSelector } from 'reselect';

import { uiSelector, autocompleteSearchesSelector, autocompleteSelector } from './state';

export const searchPathSelector = createSelector(
  uiSelector,
  ui => ui.search.path
);

const selectedAutocompleteId = createSelector(
  uiSelector,
  ui => ui.autocomplete.selectedSuggestionId
);

export const selectedFiltersSelector = createSelector(
  uiSelector,
  ui => ui.search.filters
);

export const suggestionsSelector = createSelector(
  autocompleteSearchesSelector,
  autocompleteSelector,
  (suggestionIds, autocomplete) => suggestionIds.map(id => autocomplete[id]!)
);

export const selectedSuggestionSelector = createSelector(
  autocompleteSelector,
  selectedAutocompleteId,
  (suggestions, suggestionId) =>
    suggestionId !== undefined ? suggestions[suggestionId] : undefined
);

export const selectedPropertyPinSelector = createSelector(
  uiSelector,
  ui => ui.map.selectedPropertyPin
);

export const highlightedPropertySelector = createSelector(
  uiSelector,
  ui => ui.list.highlightedProperty
);
