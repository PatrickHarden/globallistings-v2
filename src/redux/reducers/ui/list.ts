import { ActionType } from 'typesafe-actions';

import { DeepPartial } from '../../../types/common/deep-partial';
import { List } from '../../../types/ui';

import * as actions from '../../actions/list';

type RenderActionTypes = ActionType<typeof actions>;

const initialState: DeepPartial<List> = {};

export default (state = initialState, action: RenderActionTypes) => {
  switch (action.type) {
    case 'LIST_PROPERTY_MOUSEENTER':
      return { ...state, highlightedProperty: action.payload };
    case 'LIST_PROPERTY_MOUSELEAVE':
      return { ...state, highlightedProperty: undefined };
    default:
      return state;
  }
};
