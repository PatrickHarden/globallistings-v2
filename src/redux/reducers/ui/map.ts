import { ActionType } from 'typesafe-actions';

import { DeepPartial } from '../../../types/common/deep-partial';
import { Map } from '../../../types/ui';

import { onPinClicked } from '../../actions/map';

type RenderActionTypes = ActionType<typeof onPinClicked>;

const initialState: DeepPartial<Map> = {};

export default (state = initialState, action: RenderActionTypes) => {
  switch (action.type) {
    case 'MAP_PIN_CLICKED':
      return { ...state, selectedPropertyPin: action.payload };
    default:
      return state;
  }
};
