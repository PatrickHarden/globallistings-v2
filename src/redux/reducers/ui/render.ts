import { ActionType } from 'typesafe-actions';

import { DeepPartial } from '../../../types/common/deep-partial';
import { Render } from '../../../types/ui';

import { renderCompleteAction } from '../../actions/render';

type RenderActionTypes = ActionType<typeof renderCompleteAction>;

const initialState: DeepPartial<Render> = {};

export default (state = initialState, action: RenderActionTypes): DeepPartial<Render> => {
  switch (action.type) {
    case 'RENDER_COMPLETE':
      return { ...state, complete: true };
    default:
      return state;
  }
};
