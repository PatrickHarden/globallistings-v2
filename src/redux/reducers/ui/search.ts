import { DeepPartial } from '../../../types/common/deep-partial';
import { Search } from '../../../types/ui';
import * as actions from '../../actions/search';
import { ActionType } from 'typesafe-actions';

export type Actions = ActionType<typeof actions>;

export default (state: DeepPartial<Search> = {}, action: Actions) => {
  switch (action.type) {
    case 'SET_SEARCH_LOCATION':
      return {
        ...state,
        path: action.payload,
      };
    case 'CHANGE_SEARCH_FILTER':
      return {
        ...state,
        filters: { ...state.filters, ...action.payload },
      };
    default:
      return state;
  }
};
