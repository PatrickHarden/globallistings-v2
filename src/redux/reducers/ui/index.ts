import { combineReducers } from 'redux';
import search from './search';
import autocomplete from './autocomplete';
import render from './render';
import map from './map';
import list from './list';

export default combineReducers({ search, autocomplete, render, map, list });
