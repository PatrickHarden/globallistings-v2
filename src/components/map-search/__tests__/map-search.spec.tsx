import React from 'react';
import { shallow } from 'enzyme';

import { FilterSelectItem } from '../../../types/ui';
import MapSearch from '../map-search';

describe('components', () => {
  describe('map-search', () => {
    const props = {
      initialSearchLocation: 'initialSearchLocation',
      placeholder: 'placeholder',
      suggestions: [{ label: 'foo', value: 'bar' }],
      handleSelect: (selection: FilterSelectItem) => selection,
      handleInput: (searchTerm: string) => searchTerm,
    };

    it('should render correctly', () => {
      expect(shallow(<MapSearch {...props} />)).toMatchSnapshot();
    });
  });
});
