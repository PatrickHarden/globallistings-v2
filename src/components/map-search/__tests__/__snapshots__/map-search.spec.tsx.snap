// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components map-search should render correctly 1`] = `
<Select
  autocomplete={true}
  handleInput={[Function]}
  handleSelect={[Function]}
  highlightCurrentValue={true}
  icon={<MagnifyingGlass />}
  initialInputValue="initialSearchLocation"
  items={
    Array [
      Object {
        "label": "foo",
        "value": "bar",
      },
    ]
  }
  placeholder="placeholder"
  sponsor={
    <img
      src="powered-by-google.png"
    />
  }
/>
`;
